doc: more docs for from_parent (#6680)
pytest-dev/pytest: doc/en/deprecations.rst

Node Construction changed to ``Node.from_parent``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. deprecated:: 5.4

> The construction of nodes new should use the named constructor ``from_parent``.

Reviewer: ```suggestion
The construction of new nodes has to use the named constructor ``from_parent``.
```

This limitation in api surface intends to enable better/simpler refactoring of the collection tree.

> This means that instead of :code:``MyItem(name="foo", parent=collector, obj=42)``

Reviewer: ```suggestion
This means that instead of :code:`MyItem(name="foo", parent=collector, obj=42)`
```

> one now has to invoke :code:``MyItem.from_parent(collector, name="foo")``.

Reviewer: ```suggestion
one now has to invoke :code:`MyItem.from_parent(collector, name="foo")`.
```

> 

Reviewer: Could also have an example (or mention) that you have to use keyword arguments now, i.e. with mypy it has to look like this:
```diff
@ mypy/test/data.py:538 @ def split_test_cases(parent: 'DataSuiteCollector', suite: 'DataSuite',
     for i in range(1, len(cases), 6):
         name, writescache, only_when, platform_flag, skip, data = cases[i:i + 6]
         platform = platform_flag[1:] if platform_flag else None
-        yield DataDrivenTestCase(parent, suite, file,
+        yield DataDrivenTestCase.from_parent(parent, suite=suite, file=file,
                                  name=add_test_name_suffix(name, suite.test_name_suffix),
                                  writescache=bool(writescache),
                                  only_when=only_when,
```

Author: done, thanks

.. deprecated:: 5.4

The construction of nodes now should use the named constructor ``from_parent``.
This limitation in api surface intends to enable better/simpler refactoring of the collection tree.

This means that instead of :code:`MyItem(name="foo", parent=collector, obj=42)`
one now has to invoke :code:`MyItem.from_parent(collector, name="foo")`.

Plugins that wish to support older versions of pytest and suppress the warning can use
`hasattr` to check if `from_parent` exists in that version:

.. code-block:: python

    def pytest_pycollect_makeitem(collector, name, obj):
        if hasattr(MyItem, "from_parent"):
            item = MyItem.from_parent(collector, name="foo")

>             item.obj = 42

Reviewer 2: This seems a bit strange to me... how are users supposed to write their constructors?

```python
class MyItem(Node):

    def __init__(self, parent, name):
        super().__init__(parent=parent, name=name)
        self.obj = None

# constructed like this:
item = MyItem.from_parent(collector, name="foo")
item.obj = 42
```

Or are they supposed to override `from_parent`?

```python
class MyItem(Node):

    def __init__(self, parent, name, obj):
        super().__init__(parent=parent, name=name)
        self.obj = obj

    @classmethod
    def from_parent(cls, parent, name, obj):
        return super().from_parent(parent=parent, name=name, obj=obj) 

# constructed like this:
item = MyItem.from_parent(collector, name="foo", obj=42)
```

?

Author: for initially users will have to override both, we will eventually migrate to a model where they only override from_parent

Reviewer 2: > for initially users will have to override both

So I assume you mean the latter sample, right?

> we will eventually migrate to a model where they only override from_parent

But how is that possible (to only override `from_parent`)? Users still need to define their attributes somewhere. 

How to you see this playing out in the future? 

Also why users even need to implement `from_parent` at all, given that the framework doesn't call that function? Seems weird to have as prerequisite to override `__init__` (which is usually OK) and then `from_parent`, which pytest won't call at all: the user is responsible for creating nodes on the impl of the appropriate hooks (`pytest_pycollect_makeitem`).

Reviewer 2: Should we update the suggestion to go for `attr.s` directly, and implementing their own `from_parent` class?

Reviewer 2: Nevermind, I realize this won't work with `attr.s` because of the superclass relationship.

> The construction of nodes new should use the named constructor ``from_parent``.

Reviewer 2: The version above should be `5.4`
